### PR TITLE
New fields altered account

### PR DESCRIPTION
**Pull request by miiu96**
## Reasoning behind the pull request

- Added extra fields in the altered accounts structure that are needed on the elastic indexer side
- Updated `elrond-go-core` module

## Pre-requisites

Based on the [Contributing Guidelines](CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by bogdan-rosianu — DISMISSED** [review]

File: outport/process/alteredaccounts/alteredAccountsProvider.go [hunk 1]
```diff
@@ -242,10 +249,14 @@ func (aap *alteredAccountsProvider) addAddressWithBalanceChangeInMap(
 
 	_, addressAlreadySelected := markedAlteredAccounts[string(address)]
 	if addressAlreadySelected {
+		markedAlteredAccounts[string(address)].isSender = markedAlteredAccounts[string(address)].isSender || isSender
```
Comment (mariusmihaic) [inline outport/process/alteredaccounts/alteredAccountsProvider.go:252]: Please extract `string(address)` in a temp variable `addressStr` and also use it below on `L256`

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go [hunk 2]
```diff
@@ -530,13 +535,19 @@ func testExtractAlteredAccountsFromPoolShouldNotIncludeReceiverAddressIfNftCreat
 	aap, _ := NewAlteredAccountsProvider(args)
 
 	res, err := aap.ExtractAlteredAccountsFromPool(&outportcore.Pool{
+		Txs: map[string]data.TransactionHandlerWithGasUsedAndFee{
+			"hh": outportcore.NewTransactionHandlerWithGasAndFee(&transaction.Transaction{
+				SndAddr: []byte("sender in shard 0 - tx 1  "),
```
Comment (mariusmihaic) [inline outport/process/alteredaccounts/alteredAccountsProvider_test.go:540]: Please extract `sender in shard 0 - tx 1` as well as its hex encoding(senderInHex = hex.EncodeTostring(...) = `73656e64657220696e2073686172642030202d20747820312020`) in separate variables and use them below.
Otherwise, it's hard to follow that the hex encoding of the above string is correct or not

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go [hunk 3]
```diff
@@ -517,6 +521,7 @@ func testExtractAlteredAccountsFromPoolShouldNotIncludeReceiverAddressIfNftCreat
 		},
 	}
 	args := getMockArgs()
+	args.AddressConverter = testscommon.NewPubkeyConverterMock(26)
```
Comment (mariusmihaic) [inline outport/process/alteredaccounts/alteredAccountsProvider_test.go:524]: If this is the len of `"sender in shard 0 - tx 1 "`, then please use it as len(variableName)

File: outport/process/alteredaccounts/alteredAccountsProvider.go [hunk 4]
```diff
@@ -242,10 +249,14 @@ func (aap *alteredAccountsProvider) addAddressWithBalanceChangeInMap(
...
 		return
 	}
 
-	markedAlteredAccounts[string(address)] = &markedAlteredAccount{}
+	markedAlteredAccounts[string(address)] = &markedAlteredAccount{
+		isSender:       isSender,
+		balanceChanged: true,
```
Comment (mariusmihaic) [inline outport/process/alteredaccounts/alteredAccountsProvider.go:258]: This flag should not be changed when `sender == receiver` and `txType!=process.MoveBalance`

Comment (mariusmihaic) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:258]: Also, what happens for self operations?
I think you end up overwriting this value, because this use-case passes both these conditions:

```
		if senderShardID == selfShardID && len(senderAddress) > 0 {
			aap.addAddressWithBalanceChangeInMap(senderAddress, markedAlteredAccounts, true)
		}
		if txType != process.InvalidTransaction && receiverShardID == selfShardID && len(receiverAddress) > 0 {
			aap.addAddressWithBalanceChangeInMap(receiverAddress, markedAlteredAccounts, false)
		}
```

File: outport/process/alteredaccounts/alteredAccountsProvider.go (hunk 1 above)
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:252]: done

File: outport/process/alteredaccounts/alteredAccountsProvider.go (hunk 4 above)
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:258]: changed the logic

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go (hunk 2 above)
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider_test.go:556]: done

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go (hunk 3 above)
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider_test.go:540]: done

File: outport/process/alteredaccounts/alteredAccountsProvider.go [hunk 5]
```diff
@@ -231,7 +231,9 @@ func (aap *alteredAccountsProvider) extractAddressesFromTxsHandlers(
 		if senderShardID == selfShardID && len(senderAddress) > 0 {
 			aap.addAddressWithBalanceChangeInMap(senderAddress, markedAlteredAccounts, true)
 		}
-		if txType != process.InvalidTransaction && receiverShardID == selfShardID && len(receiverAddress) > 0 {
+		balanceChanged := txHandler.GetValue().Cmp(big.NewInt(0)) > 0
+		addReceiver := txType != process.InvalidTransaction && receiverShardID == selfShardID && len(receiverAddress) > 0 && balanceChanged
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:235]: done

File: outport/process/alteredaccounts/alteredAccountsProvider.go [hunk 6]
```diff
@@ -247,13 +249,15 @@ func (aap *alteredAccountsProvider) addAddressWithBalanceChangeInMap(
 		return
 	}
 
-	_, addressAlreadySelected := markedAlteredAccounts[string(address)]
+	adrStr := string(address)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:252]: don;t like this name

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go [hunk 7]
```diff
@@ -830,6 +832,112 @@ func testExtractAlteredAccountsFromPoolAddressHasMultipleNfts(t *testing.T) {
...
+		MetaData:   nil,
+	}, res[encodedAddrSnd].Tokens[0])
+
+	encodedAddrRcv := args.AddressConverter.Encode([]byte("rcv"))
+	require.False(t, res[encodedAddrRcv].BalanceChanged)
+	require.False(t, res[encodedAddrRcv].IsSender)
+	require.Len(t, res[encodedAddrRcv].Tokens, 1)
+}
+
+func testExtractAlteredAccountsFromPoolReceiverShouldHaveBalanceChanged(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider_test.go:920]: added

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go [hunk 8]
```diff
@@ -830,6 +832,112 @@ func testExtractAlteredAccountsFromPoolAddressHasMultipleNfts(t *testing.T) {
...
+					},
+				},
+			},
+		},
+	}, shared.AlteredAccountsOptions{})
+	require.NoError(t, err)
+
+	encodedAddrSnd := args.AddressConverter.Encode([]byte("snd"))
+
+	require.Len(t, res, 2)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider_test.go:902]: done

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go [hunk 9]
```diff
@@ -830,6 +832,112 @@ func testExtractAlteredAccountsFromPoolAddressHasMultipleNfts(t *testing.T) {
...
+		Balance:    expectedToken.Value.String(),
+		Nonce:      0,
+		Properties: "ok",
+		MetaData:   nil,
+	}, res[encodedAddrSnd].Tokens[0])
+
+	encodedAddrRcv := args.AddressConverter.Encode([]byte("rcv"))
+	require.False(t, res[encodedAddrRcv].BalanceChanged)
+	require.False(t, res[encodedAddrRcv].IsSender)
+	require.Len(t, res[encodedAddrRcv].Tokens, 1)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider_test.go:917]: done.

File: outport/process/alteredaccounts/alteredAccountsProvider_test.go [hunk 10]
```diff
@@ -96,6 +96,8 @@ func TestAlteredAccountsProvider_ExtractAlteredAccountsFromPool(t *testing.T) {
 	t.Run("should include receiver from tokens logs", testExtractAlteredAccountsFromPoolShouldIncludeDestinationFromTokensLogsTopics)
 	t.Run("should work when an address has balance changes, esdt and nft", testExtractAlteredAccountsFromPoolAddressHasBalanceChangeEsdtAndfNft)
 	t.Run("should work when an address has multiple nfts with different nonces", testExtractAlteredAccountsFromPoolAddressHasMultipleNfts)
+	t.Run("should not return balanceChanged for a receiver on an ESDTTransfer", testExtractAlteredAccountsFromPoolESDTTransferBalanceNotChanged)
+	t.Run("should return balanceChanged for sender and receiver", testExtractAlteredAccountsFromPoolReceiverShouldHaveBalanceChanged)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider_test.go:100]: done.

File: outport/process/alteredaccounts/alteredAccountsProvider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:263]: `addressStr` ?

File: outport/process/alteredaccounts/shared/shared.go [hunk 11]
```diff
@@ -14,9 +14,10 @@ var errInvalidAlteredAccountsOptions = errors.New("invalid altered accounts opti
 
 // AlteredAccountsOptions holds some configurable parameters to be used for extracting the altered accounts
 type AlteredAccountsOptions struct {
-	WithCustomAccountsRepository bool
-	AccountsRepository           state.AccountsRepository
-	AccountQueryOptions          api.AccountQueryOptions
+	WithCustomAccountsRepository  bool
+	AdditionalAlteredAccountsData bool
```
Comment (bogdan-rosianu) [inline outport/process/alteredaccounts/shared/shared.go:18]: might rename to `WithAdditionalOutportData`

**Review by mariusmihaic — COMMENTED** [review]
No new comments, only last 2

File: outport/process/alteredaccounts/alteredAccountsProvider.go [hunk 12]
```diff
@@ -231,7 +231,9 @@ func (aap *alteredAccountsProvider) extractAddressesFromTxsHandlers(
 		if senderShardID == selfShardID && len(senderAddress) > 0 {
 			aap.addAddressWithBalanceChangeInMap(senderAddress, markedAlteredAccounts, true)
 		}
-		if txType != process.InvalidTransaction && receiverShardID == selfShardID && len(receiverAddress) > 0 {
+		balanceChanged := txHandler.GetValue().Cmp(big.NewInt(0)) > 0
```
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:243]: done

File: outport/process/alteredaccounts/alteredAccountsProvider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/alteredaccounts/alteredAccountsProvider.go:263]: renamed

File: outport/process/alteredaccounts/shared/shared.go (hunk 11 above)
Comment (miiu96) [reply outport/process/alteredaccounts/shared/shared.go:18]: done

**Review by bogdan-rosianu — APPROVED** [review]

**Review by mariusmihaic — APPROVED** [review]
